.tools/create_algo_selection_code.py

```diff
@@ -0,0 +1,415 @@
+import importlib
+import inspect
+import pkgutil
+import textwrap
+from itertools import combinations
+from types import ModuleType
+from typing import Callable, Type
+
+from optimagic.config import OPTIMAGIC_ROOT
+from optimagic.optimization.algorithm import Algorithm
+from optimagic.typing import AggregationLevel
+
+
+def main():
+    """Create the source code for algorithms.py.
+
+    The main part are nested dataclasses that enable filtered autocomplete for algorithm
+    selection.
+
+    """
```

Comment:
I would like to see a summary of how the code generation works in the docstring. Like, with what do we start, what do we want, and what are the steps to get there.